docs/learn/multi-format-reader.md

```diff
@@ -0,0 +1,263 @@
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.extensions = {
+            ".yml": self.handle_eln_file,
+            ".yaml": self.handle_eln_file,
+            ".json": self.set_config_file,
+            # Here, one must add functions for handling any other file extension(s)
+        }
+# This has to be set to allow the convert script to use this reader. Set it to "MyDataReader".
+READER = MyDataReader
+```
+
+In order to understand the capabilities of the `MultiFormatReader` and which methods need to be implemented when extending it, we will have a look at its ```read``` method:
+```python
+def read(
+    self,
+    template: dict = None,
+    file_paths: Tuple[str] = None,
+    objects: Optional[Tuple[Any]] = None,
+    **kwargs,
+) -> dict:
+    self.kwargs = kwargs
+    self.config_file = self.kwargs.get("config_file", self.config_file)
+    self.overwrite_keys = self.kwargs.get("overwrite_keys", self.overwrite_keys)   
+```
+### Template initialization and processing order
+An empty `Template` object is initialized that later gets filled from the data files later.
+```python
+    template = Template(overwrite_keys=self.overwrite_keys)
+
+    def get_processing_order(path: str) -> Tuple[int, Union[str, int]]:
+        """
+        Returns the processing order of the file.
+        """
+        ext = os.path.splitext(path)[1]
+        if self.processing_order is None or ext not in self.processing_order:
+            return (1, ext)
+        return (0, self.processing_order.index(ext))
+
+    sorted_paths = sorted(file_paths, key=get_processing_order)
+```
+If the reader has a `self.processing_order`, the input files get sorted in this order. If `self.overwrite_keys` is True, later files get precedent.
+
+### Reading of input files
+```python
+    for file_path in sorted_paths:
+        extension = os.path.splitext(file_path)[1].lower()
+        if extension not in self.extensions:
+            logger.warning(
+                f"File {file_path} has an unsupported extension, ignoring file."
+            )
+            continue
+        if not os.path.exists(file_path):
+            logger.warning(f"File {file_path} does not exist, ignoring entry.")
+            continue
+
+        template.update(self.extensions.get(extension, lambda _: {})(file_path))
+```
+This parts reads in the data from all data files. The `MultiFormatReader` has an `extensions` property, which is a dictionary that for each file extension calls a function that reads in data from files with that extension. If the reader shall handle e.g. an HDF5 file, a method for handling this type of file should be added, i.e., `self.extensions[".hdf5"] = self.handle_hdf5`. 
+Note that these method should also implement any case selection logic, i.e., it may not be sufficient to rely on the filename suffix, but when may also need to check for multiple versions, binary signature, mimetype, etc.

```

Comment:
```suggestion
Note that these methods should also implement any logic depending on the provided data, i.e., it may not be sufficient to rely on the filename suffix, but the reader may also need to check for different file versions, binary signature, mimetype, etc.

```

Reply:
changed